# Análisis Estratégico y Storytelling Visual.

In [9]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

# ==========================================
# CONFIGURACIÓN INICIAL Y ESTILO
# ==========================================
# Configuramos un estilo visual limpio y profesional para los gráficos
sns.set_theme(style="whitegrid")
plt.rcParams['figure.figsize'] = [10, 6]

print("⏳ Iniciando el proceso de análisis y transformación...")

# ==========================================
# FASE 1 & 2: CARGA Y LIMPIEZA DE DATOS (ETL)
# ==========================================
# 1. Cargar el dataset original
df = pd.read_csv("COVID_Clean.csv")


⏳ Iniciando el proceso de análisis y transformación...


/var/folders/cr/f3dqyy756z9ftxxsrmgtgjjh0000gn/T/ipykernel_37151/1273621649.py:18: DtypeWarning: Columns (6) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("COVID_Clean.csv")


In [10]:
df.columns

Index(['Tweet URL', 'Tweet Posted Time (UTC)', 'Tweet Content', 'Tweet Type',
       'Client', 'Retweets Received', 'Likes Received', 'Tweet Location',
       'Tweet Language', 'User Id', 'Name', 'Screen Name', 'User Bio',
       'Verified or Non-Verified', 'Profile URL', 'Protected or Non-protected',
       'User Followers', 'User Following', 'User Account Creation Date',
       'Año-Mes', 'Hora del Día', 'Pais_Estandarizado', 'Categoria_Influencia',
       'Tiene_Mencion', 'Tiene_Enlace', 'Dispositivo_Limpio'],
      dtype='object')

In [11]:
# ==========================================
# GENERACIÓN DE GRÁFICOS (VISUALIZACIÓN EN PYTHON)
# ==========================================
print("📊 Generando gráficos estadísticos...")

# --- GRÁFICO 1: Distribución por País Estandarizado (Excluyendo ruidos) ---
# Filtramos los datos para mostrar solo los países que logramos clasificar con éxito
paises_validos = df[~df['Pais_Estandarizado'].isin(['Otros / No Identificado', 'No Especificado'])]

plt.figure(figsize=(10, 5))
sns.countplot(
    data=paises_validos, 
    y='Pais_Estandarizado', 
    order=paises_validos['Pais_Estandarizado'].value_counts().index, 
    palette='viridis'
)
plt.title('Distribución de Tweets por País Estandarizado (Top Mercados)', fontsize=14, pad=15)
plt.xlabel('Cantidad de Tweets', fontsize=12)
plt.ylabel('País', fontsize=12)
plt.tight_layout()

# Guardamos el primer gráfico como imagen
plt.savefig('./images/tweets_por_pais.png')
plt.close()
print("💾 Gráfico 1 guardado como 'tweets_por_pais.png'")



/var/folders/cr/f3dqyy756z9ftxxsrmgtgjjh0000gn/T/ipykernel_37151/485954881.py:11: FutureWarning: 

Passing `palette` without assigning `hue` is deprecated and will be removed in v0.14.0. Assign the `y` variable to `hue` and set `legend=False` for the same effect.

  sns.countplot(


📊 Generando gráficos estadísticos...
💾 Gráfico 1 guardado como 'tweets_por_pais.png'


In [12]:
# =====================================================================
# GRÁFICO 2: Volumen de tuits por Hora del Día
# =====================================================================
plt.figure(figsize=(11, 5))
# Agrupamos por hora y contamos cuántos tuits hay en cada una
tuits_por_hora = df.groupby('Hora del Día').size().reset_index(name='Cantidad de Tweets')

sns.lineplot(data=tuits_por_hora, x='Hora del Día', y='Cantidad de Tweets', marker='o', color='#2b5c8f', linewidth=2.5)
plt.title('Ritmo de la Conversación: Volumen de Tweets por Hora del Día (UTC)', fontsize=13, pad=15)
plt.xlabel('Hora del Día (Formato 24h)', fontsize=11)
plt.ylabel('Cantidad de Tweets', fontsize=11)
plt.xticks(range(0, 24))  # Forzamos a mostrar todas las horas del día
plt.tight_layout()
plt.savefig('./images/analisis_horas_pico.png')
plt.close()
print("💾 Guardado: 'analisis_horas_pico.png'")



💾 Guardado: 'analisis_horas_pico.png'


In [13]:

# =====================================================================
# GRÁFICO 4: Dispositivos favoritos por Idioma (Top 2 Idiomas)
# =====================================================================
plt.figure(figsize=(10, 6))
# Filtramos solo para los dos idiomas mayoritarios que descubrimos en la Fase 1
top_idiomas = df[df['Tweet Language'].isin(['English', 'Spanish'])]

# Creamos una tabla cruzada normalizada al 100% para comparar proporciones de forma justa
tabla_dispositivos = pd.crosstab(top_idiomas['Tweet Language'], top_idiomas['Dispositivo_Limpio'], normalize='index') * 100

tabla_dispositivos.plot(kind='barh', stacked=True, figsize=(10, 5), colormap='plasma')
plt.title('Perfil Tecnológico: Dispositivo Utilizado según el Idioma del Usuario', fontsize=13, pad=15)
plt.xlabel('Porcentaje (%) de Usuarios', fontsize=11)
plt.ylabel('Idioma', fontsize=11)
plt.legend(title='Dispositivo', bbox_to_anchor=(1.05, 1), loc='upper left')
plt.tight_layout()
plt.savefig('./images/dispositivos_por_idioma.png')
plt.close()
print("💾 Guardado: 'dispositivos_por_idioma.png'")


💾 Guardado: 'dispositivos_por_idioma.png'


<Figure size 1000x600 with 0 Axes>

In [14]:
df.columns

Index(['Tweet URL', 'Tweet Posted Time (UTC)', 'Tweet Content', 'Tweet Type',
       'Client', 'Retweets Received', 'Likes Received', 'Tweet Location',
       'Tweet Language', 'User Id', 'Name', 'Screen Name', 'User Bio',
       'Verified or Non-Verified', 'Profile URL', 'Protected or Non-protected',
       'User Followers', 'User Following', 'User Account Creation Date',
       'Año-Mes', 'Hora del Día', 'Pais_Estandarizado', 'Categoria_Influencia',
       'Tiene_Mencion', 'Tiene_Enlace', 'Dispositivo_Limpio'],
      dtype='object')

In [22]:
df_agrupado = df.groupby('Categoria_Influencia')['Retweets Received'].mean().reset_index()

# Ordenamos manualmente las categorías para que la gráfica tenga sentido lógico en la pizarra
orden_logico = [
    'Micro-usuario (0-500)', 
    'Usuario Activo (501-5k)', 
    'Micro-Influencer (5k-50k)', 
    'Celebridad / Institución (>50k)'
]
# Seteamos la columna como categoría con orden para que Matplotlib lo respete
df_agrupado['Categoria_Influencia'] = pd.Categorical(df_agrupado['Categoria_Influencia'], categories=orden_logico, ordered=True)
df_agrupado = df_agrupado.sort_values('Categoria_Influencia')

# =====================================================================
# DISEÑO DEL GRÁFICO CON MATPLOTLIB PURO
# =====================================================================
fig, ax = plt.subplots(figsize=(10, 6))

# Definimos colores corporativos elegantes (Gris para la masa, Azul/Rojo para el pico)
colores = ['#bdc3c7', '#95a5a6', '#34495e', '#e74c3c']

# Dibujamos las barras
barras = ax.bar(
    df_agrupado['Categoria_Influencia'], 
    df_agrupado['Retweets Received'], 
    color=colores, 
    edgecolor='none',
    width=0.6
)

# Añadimos títulos y etiquetas con formato limpio para la audiencia
ax.set_title('El Impacto de la Audiencia: Promedio de Retweets por Tipo de Cuenta', fontsize=14, pad=20, fontweight='bold', color='#2c3e50')
ax.set_xlabel('Categoría del Usuario (Basada en Seguidores)', fontsize=11, labelpad=12)
ax.set_ylabel('Promedio de Retweets Recibidos', fontsize=11, labelpad=12)

# Añadimos etiquetas de datos encima de cada barra para que sea autoexplicativo
for barra in barras:
    alto = barra.get_height()
    ax.annotate(
        f'{alto:.1f}',  # Muestra el promedio con un solo decimal
        xy=(barra.get_x() + barra.get_width() / 2, alto),
        xytext=(0, 3),  # 3 puntos de desfase vertical hacia arriba
        textcoords="offset points",
        ha='center', 
        va='bottom', 
        fontsize=10, 
        fontweight='bold'
    )

# Estética final: Limpiamos las líneas superiores y derechas (Spines) para un look moderno
ax.spines['top'].set_visible(False)
ax.spines['right'].set_visible(False)
ax.spines['left'].set_color('#cccccc')
ax.spines['bottom'].set_color('#cccccc')

# Añadimos una cuadrícula sutil solo en el eje Y
ax.yaxis.grid(True, linestyle='--', alpha=0.5, color='#cccccc')
ax.set_axisbelow(True) # Hace que la cuadrícula se pinte por detrás de las barras

plt.tight_layout()

# Guardamos la imagen final
plt.savefig('./images/grafico_influencia_matplotlib.png', dpi=300)
plt.close()

print("💾 ¡Gráfico de influencia generado con éxito en './images/grafico_influencia_matplotlib.png'!")

💾 ¡Gráfico de influencia generado con éxito en './images/grafico_influencia_matplotlib.png'!


Al calcular los promedios, queda en evidencia una asimetría brutal. Mientras que los millones de "Micro-usuarios" (con menos de 500 seguidores) apenas rozan un promedio cercano a 0 interacciones por tuit, las cuentas categorizadas como "Celebridad / Institución" (con más de 50,000 seguidores) disparan el promedio de retweets de forma exponencial. 

Esto demuestra un comportamiento clásico de "Ley de Potencia" o "Efecto Mateo" en la comunicación digital: el que más tiene, más recibe. La viralidad durante la crisis de la COVID-19 no fue democrática. La capacidad de instalar una narrativa, propagar un miedo o difundir una calma estuvo concentrada en un puñado muy pequeño de cuentas de gran tamaño.

Si una empresa o gobierno quiere mitigar la desinformación (Fake News) o lanzar una campaña de concientización masiva, el esfuerzo analítico no debe diluirse escuchando a toda la masa por igual. El foco debe ser un monitoreo quirúrgico de las cuentas del segmento "Celebridad", ya que un solo tuit de este grupo tiene más impacto que miles de tuits de usuarios comunes juntos.

En los gráficos se observa un "escalón" muy interesante en el segmento intermedio. Los usuarios activos y los micro-influencers empiezan a levantar tracción y a superar el promedio base de retweets.

Durante una crisis de salud pública, las grandes instituciones (Celebridades) emiten los datos oficiales, pero los "Micro-Influencers" (médicos divulgadores, científicos locales, periodistas sectoriales) actúan como traductores de la información. Aunque no tienen millones de seguidores, su promedio de retweets es saludable porque su comunidad es altamente cohesionada y confía ciegamente en su criterio profesional.

## Tienen mencion o enlace

In [17]:
total = len(df)
menciones = (df['Tiene_Mencion'] == 'Sí').sum()
enlaces = (df['Tiene_Enlace'] == 'Sí').sum()
porc_menciones = (menciones / total) * 100
porc_enlaces = (enlaces / total) * 100

print(f"📊 Resumen de Contenido Operacional:")
print(f"-> Total del Dataset: {total} tuits")
print(f"-> Con Menciones (@): {menciones} tuits ({porc_menciones:.2f}%)")
print(f"-> Con Enlaces (http): {enlaces} tuits ({porc_enlaces:.2f}%)")

📊 Resumen de Contenido Operacional:
-> Total del Dataset: 60163 tuits
-> Con Menciones (@): 49029 tuits (81.49%)
-> Con Enlaces (http): 50120 tuits (83.31%)


In [18]:
fig, ax = plt.subplots(figsize=(8, 5))

# Variables para el gráfico
metricas = ['Contiene Enlaces (http)', 'Contiene Menciones (@)']
porcentajes = [porc_enlaces, porc_menciones]
colores = ['#2ecc71', '#3498db']  # Verde para enlaces (fuentes), Azul para menciones (interacción)

# Dibujar las barras horizontales para una lectura más cómoda
barras = ax.barh(metricas, porcentajes, color=colores, height=0.5, edgecolor='none')

# Configuración de títulos e insights ejecutivos
ax.set_title('Estructura Operacional del Mensaje: Análisis de Contenido', fontsize=14, pad=20, fontweight='bold', color='#2c3e50')
ax.set_xlabel('Porcentaje del Total de Tweets (%)', fontsize=11, labelpad=10)
ax.set_xlim(0, 100) # Fijamos el límite al 100% para dar perspectiva real

# Añadir las etiquetas de datos exactas al final de cada barra
for barra in barras:
    ancho = barra.get_width()
    ax.annotate(
        f'{ancho:.1f}%',  # Muestra el porcentaje con un decimal
        xy=(ancho, barra.get_y() + barra.get_height() / 2),
        xytext=(10, 0),   # Desfase de 10 puntos a la derecha de la barra
        textcoords="offset points",
        ha='left', 
        va='center', 
        fontsize=11, 
        fontweight='bold',
        color='#2c3e50'
    )

# Estética y limpieza visual: Quitamos los bordes innecesarios
ax.spines['top'].set_visible(False)
ax.spines['right'].set_visible(False)
ax.spines['bottom'].set_color('#cccccc')
ax.spines['left'].set_color('#cccccc')

# Añadimos líneas de cuadrícula verticales sutiles para guiar la vista
ax.xaxis.grid(True, linestyle='--', alpha=0.5, color='#cccccc')
ax.set_axisbelow(True)

plt.tight_layout()

# Guardar la imagen en tu carpeta de VS Code
plt.savefig('./images/analisis_estructura_tweets.png', dpi=300)
plt.close()

print("💾 ¡Gráfico guardado con éxito en 'analisis_estructura_tweets.png'!")
print(f"Resultados validados: Enlaces = {porc_enlaces:.2f}%, Menciones = {porc_menciones:.2f}%")

💾 ¡Gráfico guardado con éxito en 'analisis_estructura_tweets.png'!
Resultados validados: Enlaces = 83.31%, Menciones = 81.49%


Tradicionalmente, las redes sociales se asocian con opiniones rápidas, rumores o contenido subjetivo. Sin embargo, un volumen del 83.3% de tuits con enlaces externos demuestra un comportamiento anómalo y maduro de la comunidad digital durante la crisis de la COVID-19. Ante una situación de incertidumbre global y crisis sanitaria, la audiencia sintió la necesidad imperativa de respaldar y legitimar sus afirmaciones. Los usuarios no solo daban su opinión, sino que adjuntaban evidencias: artículos periodísticos, hilos científicos, boletines oficiales del Estado o reportes de la OMS.

La conversación estuvo altamente interconectada y dirigida. Los usuarios utilizaron activamente el símbolo @ para interactuar

Al combinar ambas métricas, descubrimos que el tuit estándar de la crisis de la COVID-19 combinaba reputación y dirección. La comunidad procesaba la información compartiendo un enlace y abriendo un debate etiquetando a terceros.